src/peer/socket.rs

```diff
@@ -99,103 +91,107 @@ async fn retainer<S: Signer>(
+                        break 'receive
+                    }
 
-                    // we take a note with when a message was received
-                    last = Instant::now();
-                    log::trace!("received a message from relay {:?}", &u.domain());
                     let message = match message {
-                        Ok(Message::Pong(_)) => {
-                            log::trace!("received a pong");
-                            continue 'receive;
-                        }
-                        Ok(message) => message,
-                        Err(err) => {
-                            // todo: those errors probably mean we need to re-connect
-                            // we will see what to do later.
-                            log::error!("disconnected: error while receiving message: {}", err);
-                            break 'receive;
+                        Ok(Some(message)) => message,
+                        Ok(None) => {
+                            //weird why would we receive a non message
+                            log::error!("socket closed!");
+                            handler.abort();
+                            return
+                        },
+                        Err(_) => {
+                            // receive timeout (on upstream message)
+                            // we can then send a ping to keep the connection alive
+                            log::debug!("sending a ping");
+                            Message::Ping(Vec::default())
                         }
                     };
-
-                    if let Err(err) = down_tx.send(message).await {
-                        log::error!("failed to queue received message for processing: {}", err);
+                    log::trace!("sending message to relay {:?}", &u.domain());
+                    if let Err(err) = write.send(message).await {
+                        // probably connection closed as well, we need to renew!
+                        log::error!("disconnected: error while sending message: {}", err);
+                        break 'receive;
                     }
                 }
+
             }
         }
+        // make sure read routine is fully closed
+        handler.abort();
         tokio::time::sleep(Duration::from_secs(2)).await;
     }
 }
 
-fn read_stream(
-    mut stream: futures_util::stream::SplitStream<
-        tokio_tungstenite::WebSocketStream<
-            tokio_tungstenite::MaybeTlsStream<tokio::net::TcpStream>,
-        >,
-    >,
-) -> mpsc::Receiver<Result<Message, Error>> {
-    let (sender, receiver) = mpsc::channel(1);
-    tokio::spawn(async move {
-        loop {
-            match stream.next().await {
-                None => return,
-                Some(result) => {
-                    if sender.send(result).await.is_err() {
-                        return;
-                    }
-                }
-            }
-        }
-    });
-
-    receiver
+fn timestamp() -> u64 {
```

Comment:
can we store instead Instant instance ?
- you won't need to unwrap here
- it is more reliable and
- you can call elapsed method on it directly to know if you passed the timeout threshold.

Reply:
No, because i need to use AtomicU64. Since the value is modified/read from 2 different routines. If i have to use Instance then i have to use a Mutex which is heavier for that use case

Reply:
I understand this but not sure if practically performance will be an issue here.
the thing is that i thought before that this method is unlikely to fail, till i saw it randomly failing and panicking. i remember i switched to instant for that reason.

should we do  at lest some error handling / logging here to spot this when if happens?

Reply:
if you check the docs, this can only fail if the duration_since is given time AFTER the time instance (or now) in this case which is impossible since i use the EPOCH.

If suddenly the EPOCH is after the now, i would crash instead